editor/gui/editor_quick_open_dialog.cpp

```diff
@@ -45,6 +46,49 @@
 #include "scene/gui/texture_rect.h"
 #include "scene/gui/tree.h"
 
+Rect2i HighlightedLabel::get_substr_rect(const Vector2i &p_substr) {
+	Ref<Font> font = get_theme_font(SceneStringName(font));
+	int font_size = get_theme_font_size(SceneStringName(font_size));
+	Vector2i prefix = font->get_string_size(get_text().substr(0, p_substr.x), HORIZONTAL_ALIGNMENT_LEFT, -1, font_size);
+	prefix.y = 0;
+	Vector2i size = font->get_string_size(get_text().substr(p_substr.x, p_substr.y), HORIZONTAL_ALIGNMENT_LEFT, -1, font_size);
+	return { prefix, size };
```

Comment:
This won't work with right-to-left or mixed direction text. Also, a substring is not necessarily continuous and can consist of multiple rects.

`TextServer.shaped_text_get_selection` should be used instead.

Reply:
Since it's a class derived from `Label`, it probably should use existing `text_rid` (currently private, but getter can be added).

```c++

// Assuming `RID get_text_rid() const { return text_rid; }` was added to `Label`.
// You can use something like this to draw rects (can be done directly from NOTIFICATION_DRAW or cached).

void draw_subst_rect(const RID &p_ci, const Vector2 &p_osf, const Vector2i &p_substr, const Color &p_color) {
	Vector<Vector2> ranges = TS->shaped_text_get_selection(get_text_rid(), p_substr.x, p_substr.y - p_substr.x); // Get horizontal ranges (start, end offsets for each).
	Size2 line_size = TS->shaped_text_get_size(get_text_rid()); // Get string size (for rect height).
	for (int i = 0; i < ranges.size(); i++) {
		Rect2 rect = Rect2(ranges[i].x + p_ofs.x, p_ofs.y, ranges[i].y - ranges[i].x, line_size.y);
		RenderingServer::get_singleton()->canvas_item_add_rect(p_ci, rect, p_color); // Draw rect.
	}
}
```

Reply:
Note: this code will only work (which should be fine for this case), for multiple lines the same can be done using `lines_rid` vector from the `Label` (just do the same for each).

Reply:
Thanks! I realized late last night I had entirely neglected RTL..